[#85] As a developer, I can create OAuth client via API

carryall opened the pull request:
Resolved https://github.com/carryall/go-google-scraper-challenge/issues/85

## What happened 👀

- [x] Move database related functions to a separate package to avoid cycle import error
- [x] Add API for OAuth client so that we can create a pair of client ID and secret

## Insight 📝

N/A

## Proof Of Work 📹

Able to create the client ID and secret from POST `/oauth/clients`

![Screen Shot 2565-05-25 at 17 41 52](https://user-images.githubusercontent.com/1772999/170244315-3c997f43-e2a7-4217-bf80-5aa3dc76652c.png)

hoangmirs commented on lib/api/v1/serializers/users.go:
```diff
@@ -0,0 +1,16 @@
+package serializers
+
+type UserResponse struct {
```
We are using JSON:API for responses in almost projects, should we support it as well?
https://github.com/google/jsonapi

carryall replied:
Ahh actually I forgot to remove this, we didn't use it anywhere yet. But I agree that we should use the [`jsonapi`](https://github.com/google/jsonapi) I'll add it when we need to response with resources with ID 🙌🏼

carryall commented on lib/api/v1/forms/registration.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+type RegistrationForm struct {
+	ClientID     string `form:"client_id"`
```
I guess we cannot as the OAuth2 token endpoint only works with the `application/x-www-form-urlencoded` content type, I've try with JSON content type before but it doesn't work, some more info can be found on [this thread](https://stackoverflow.com/questions/57670209/oauth-2-0-rfc-6749-content-type-application-json-is-compliant-with-spec)

hoangmirs replied:
I see, thank you

carryall commented on lib/api/v1/forms/registration.go:
```diff
@@ -0,0 +1,65 @@
+		return false, err
+	}
+
+	emailAlreadyExist := models.UserEmailAlreadyExist(f.Email)
```
Yes, updated in [485416f](https://github.com/carryall/go-google-scraper-challenge/pull/83/commits/485416fb11e7d0c227989337c8f33575409db00e)

carryall commented on lib/api/v1/forms/registration_test.go:
```diff
@@ -0,0 +1,189 @@
+	. "go-google-scraper-challenge/test"
+
+	"github.com/bxcodec/faker/v3"
+	"github.com/onsi/ginkgo"
```
🙈  updated in [871710b](https://github.com/carryall/go-google-scraper-challenge/pull/83/commits/871710b8daf2abf93ba84ecafd3d205ed15ccee0) along with the comments below

carryall commented on lib/api/v1/routers/router.go:
```diff
@@ -9,5 +9,11 @@ import (
+	registerController := controllers.UsersController{}
+
+	v1.GET("/health", healthController.HealthStatus)
+	v1.POST("/auth_clients", authClientsController.Create)
```
updated in [be3a1c0](https://github.com/carryall/go-google-scraper-challenge/pull/83/commits/be3a1c085c8a9c347173091ecf1101c9e8e9097c)

Lahphim reviewed (commented):
I think this PR looks huge as there is more than 1 action within the ticket.
So later on we can do with 2 options here:

1. Break down the ticket to some smaller one
2. Break down the PR

carryall commented on helpers/api/error.go:
```diff
@@ -0,0 +1,18 @@
+package api_helpers
```
fixed in [b68dc75](https://github.com/carryall/go-google-scraper-challenge/pull/83/commits/b68dc7584f0c16d0c87b43d061127b074ca80095)

carryall commented:
@Lahphim Fixed by
- Split the issue to
  - #85 
  - #82 
- Split the PR into
  - This PR for #85 
  - [Another PR](https://github.com/carryall/go-google-scraper-challenge/pull/86) for #82 

@hoangmirs You might want to review it again 🙇🏼 

> I think this PR looks huge as there is more than 1 action within the ticket. So later on we can do with 2 options here:
> 
> 1. Break down the ticket to some smaller one
> 2. Break down the PR

Lahphim commented on helpers/api/error.go:
```diff
@@ -0,0 +1,17 @@
+package helpers
+
+import (
+	"fmt"
+	"go-google-scraper-challenge/lib/api/v1/serializers"
+
+	"github.com/gin-gonic/gin"
```
I used to faced this issue so many times once I add or remove in the file 😆

Lahphim reviewed (approved):
Looks great!
